osh/builtin_assign.py

```diff
@@ -25,10 +26,108 @@
   from frontend import arg_def
 
 
+def _PrintVariables(mem, cmd_val, arg, print_flags, readonly = False, exported = False):
+  flag_g = getattr(arg, 'g', None)
+  flag_n = getattr(arg, 'n', None)
+  flag_r = getattr(arg, 'r', None)
+  flag_x = getattr(arg, 'x', None)
+  flag_a = getattr(arg, 'a', None)
+  flag_A = getattr(arg, 'A', None)
+
+  lookup_mode = scope_e.Dynamic
+  if cmd_val.builtin_id == builtin_i.local:
+    if flag_g and not mem.IsGlobalScope():
+      return 1
+    lookup_mode = scope_e.LocalOnly
+  elif flag_g:
+    lookup_mode = scope_e.GlobalOnly
+
+  if len(cmd_val.pairs) == 0:
+    print_all = True
+    cells = mem.GetAllCells(lookup_mode)
+    names = sorted(cells)
+  else:
+    print_all = False
+    names = []
+    cells = {}
+    for pair in cmd_val.pairs:
+      name = pair.lval.name
+      if pair.rval and pair.rval.tag == value_e.Str:
```

Comment:
It's better to use `tag_()` though it's unfortunately uglier.  That's because it translates to C++ better -- C++ doesn't have "virtual" fields like Python, only methods.  I plan to get rid of `.tag` and make it `.tag()`, but for now it's `.tag_()`

Reply:
OK! Updated 0bd3c4a3.